### PR TITLE
add concurrency limitation for super proxy

### PR DESCRIPTION
**Pull request by daizong**

### CODE REVIEW

File: usage/usage.go [hunk 1]
```diff
@@ -29,6 +32,12 @@ func (u *ProxyUsage) Start() {
 	u.outgoingChan = make(chan uint64, DEFAULT_CHAN_CAP)
 	u.stop = make(chan struct{})
 	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				fmt.Println("Recovered in ProxyUsage.Start", r)
```
Comment (cszichao) [inline usage/usage.go]: log this error instead

File: superproxy/superproxy.go [hunk 2]
```diff
@@ -48,6 +51,9 @@ type SuperProxy struct {
 
 	//usage
 	Usage *usage.ProxyUsage
+
+	//concurrency chan
+	concurrencyChan chan struct{}
```
Comment (cszichao) [inline superproxy/superproxy.go]: why struct? any better types?

**Comment by cszichao** [comment]
add traffic counting & concurrency limit support for https decryption

File: superproxy/superproxy.go (hunk 2 above)
Comment (daizong) [reply superproxy/superproxy.go]: reference: https://dave.cheney.net/2014/03/25/the-empty-struct

File: usage/usage.go [hunk 3]
```diff
@@ -29,6 +33,12 @@ func (u *ProxyUsage) Start() {
 	u.outgoingChan = make(chan uint64, DEFAULT_CHAN_CAP)
 	u.stop = make(chan struct{})
 	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				util.ErrWrapper(nil, "recover on ProxyUsage.Start: %v", r)
```
Comment (cszichao) [inline usage/usage.go]: not a logger, just made an error

File: proxy/proxy/handler.go [hunk 4]
```diff
@@ -107,10 +114,22 @@ func (h *Handler) sendHTTPSProxyStatusBadGateway(c net.Conn) (err error) {
 	return util.WriteWithValidation(c, []byte("HTTP/1.1 501 Bad Gateway\r\n\r\n"))
 }
 
+var (
+	HttpOkByteSize         = uint64(len([]byte("HTTP/1.1 200 OK\r\n\r\n")))
```
Comment (cszichao) [inline proxy/proxy/handler.go]: const httpTunnelMade string = "HTTP/1.1 200 OK\r\n\r\n"
var httpTunnelMadeSize = balabla

Comment (cszichao) [reply proxy/proxy/handler.go]: do NOT ever made 2 copies of the same string, which made lead several unexpected bugs

**Pull request merged**